base/util.go

```diff
@@ -1056,3 +1057,70 @@ func Sha1HashString(str string, salt string) string {
 	hashedKey := h.Sum(nil)
 	return fmt.Sprintf("%x", hashedKey)
 }
+
+// InjectJSONProperty takes the given JSON byte slice, marshals val, and inserts under the given key, and returns the result.
+func InjectJSONProperty(b []byte, key string, val interface{}) (new []byte, err error) {
+	bIsJSONObject, bIsEmpty := isJSONObject(b)
+	if !bIsJSONObject {
+		return nil, errors.New("b is not a JSON object")
+	}
+
+	valBytes, err := json.Marshal(val)
+	if err != nil {
+		return nil, err
+	}
+
+	return injectJSONPropertyFromBytes(b, bIsEmpty, key, valBytes), nil
+}
+
+// InjectJSONPropertyFromBytes takes the given JSON byte slice, and inserts val under the given key, and returns the result.
+func InjectJSONPropertyFromBytes(b []byte, key string, val []byte) (new []byte, err error) {
+	bIsJSONObject, bIsEmpty := isJSONObject(b)
+	if !bIsJSONObject {
+		return nil, errors.New("b is not a JSON object")
+	}
+
+	return injectJSONPropertyFromBytes(b, bIsEmpty, key, val), nil
+}
+
+// isJSONObject checks if the given bytes are a JSON object,
+// and also whether it's an empty object or not.
+func isJSONObject(b []byte) (isJSONObject, isEmpty bool) {
+
+	// Check if the byte slice starts with { and ends with }
+	if len(b) < 2 || b[0] != 0x7b || b[len(b)-1] != 0x7d {
+		return false, false
+	}
+
+	return true, len(b) == 2
+}
+
+// injectJSONPropertyFromBytes injects val under the given key into b, assuming b is a JSON object.
+func injectJSONPropertyFromBytes(b []byte, bIsEmpty bool, key string, val []byte) (new []byte) {
+
+	// the overhead of wrapping key in quotes, a colon, and a comma if body is not empty
+	jsonOverhead := 4
+	if bIsEmpty {
+		jsonOverhead = 3
+	}
+
+	// Create the new byte slice with the required capacity
+	new = make([]byte, 0, len(b)+len(val)+len(key)+jsonOverhead)
+
+	// copy almost all of b, except the last closing brace
+	new = append(new, b[0:len(b)-1]...)
```

Comment:
Is copy more efficient than append here (and below)?

Reply:
Is marginally faster using copy (~5%) - changed

```
18:48 $ benchcmp bench-append.out bench-copy.out
benchmark                                                 old ns/op     new ns/op     delta
BenchmarkInjectJSONPropertyFromBytes/null-12              30.1          30.1          +0.00%
BenchmarkInjectJSONPropertyFromBytes/{}-12                68.1          65.1          -4.41%
BenchmarkInjectJSONPropertyFromBytes/{"key":"val"}-12     73.3          69.7          -4.91%
BenchmarkInjectJSONProperty/null-12                       40.1          39.9          -0.50%
BenchmarkInjectJSONProperty/{}-12                         228           227           -0.44%
BenchmarkInjectJSONProperty/{"key":"val"}-12              234           235           +0.43%

benchmark                                                 old allocs     new allocs     delta
BenchmarkInjectJSONPropertyFromBytes/null-12              1              1              +0.00%
BenchmarkInjectJSONPropertyFromBytes/{}-12                1              1              +0.00%
BenchmarkInjectJSONPropertyFromBytes/{"key":"val"}-12     1              1              +0.00%
BenchmarkInjectJSONProperty/null-12                       2              2              +0.00%
BenchmarkInjectJSONProperty/{}-12                         3              3              +0.00%
BenchmarkInjectJSONProperty/{"key":"val"}-12              3              3              +0.00%

benchmark                                                 old bytes     new bytes     delta
BenchmarkInjectJSONPropertyFromBytes/null-12              16            16            +0.00%
BenchmarkInjectJSONPropertyFromBytes/{}-12                64            64            +0.00%
BenchmarkInjectJSONPropertyFromBytes/{"key":"val"}-12     80            80            +0.00%
BenchmarkInjectJSONProperty/null-12                       24            24            +0.00%
BenchmarkInjectJSONProperty/{}-12                         32            32            +0.00%
BenchmarkInjectJSONProperty/{"key":"val"}-12              48            48            +0.00%
```